tests/test_bgp_af.rb

```diff
@@ -28,31 +28,254 @@ def setup
+    [:advertise_l2vpn_evpn,          :nexus,  'default', %w(ipv4 multicast),   :CliError],
+    [:advertise_l2vpn_evpn,          :nexus,  'default', %w(ipv6 multicast),   :CliError],
+    [:advertise_l2vpn_evpn,          :nexus,  'red',     %w(ipv4 multicast),   :CliError],
+    [:advertise_l2vpn_evpn,          :nexus,  'red',     %w(ipv6 multicast),   :CliError],
+  ]
+
+  # rubocop:disable Style/SpaceAroundOperators
+  def check_test_exceptions(test_, os_, vrf_, af_)
+    ret = nil
+    amb = nil
+    @@test_exceptions.each do |test, os, vrf, af, expect|
+      next unless (test_ == test || test == :any) &&
+                  (os_   == os   || os   == :any) &&
+                  (vrf_  == vrf  || vrf  == :any) &&
+                  (af_   == af   || af   == :any)
+      return expect if expect == :success || expect == :skip
+
+      # Otherwise, make sure there's no ambiguity/overlap in the exceptions.
+      assert_nil(ret, 'TEST ERROR: Exceptions matrix has ambiguous entries! ' \
+                      "#{amb} and [#{test}, #{os}, #{vrf}, #{af}]")
+      ret = expect
+      amb = [test, os, vrf, af, expect]
+    end
+    # Return the expected test result
+    ret.nil? ? :success : ret
   end
-
-  # show bgp ipv4 unicast dampening parameters
-  # Route Flap Dampening Parameters for VRF default Address family IPv4 Unicast:
-  # Default values in use:
-  # Half-life time                 : 15 mins
-  # Suppress penalty               : 2000
-  # Reuse penalty                  : 750
-  # Max suppress time              : 45 mins
-  # Max suppress penalty           : 6000
-  def get_bgp_af_dampening_params(_asn, vrf, af)
-    afi = af.first
-    safi = af.last
-    @device.cmd("show bgp vrf #{vrf} #{afi} #{safi} dampening parameters")
+  # rubocop:enable Style/SpaceAroundOperators
+
+  def test_properties_matrix
+    @@tasns.each do |asn|
+      config_ios_xr_dependencies(asn) if platform == :ios_xr
+
+      @@tvrfs.each do |vrf|
+        @@tafs.each do |af|
+          puts '**************************************'
+          @@tvalues.each do |test, test_values|
+            puts "******** #{test}, #{asn}, #{vrf}, #{af}"
+
+            # Override expectation for some specific cases..
+            expect = check_test_exceptions(test, platform, vrf, af)
+
+            # Setup
+            bgp_af = RouterBgpAF.new(asn, vrf, af)
+            initial = bgp_af.send(test)
+            default = bgp_af.send("default_#{test}")
+
+            if expect == :skip
+              # Do nothing..
+              puts '         skip'
+
+            elsif expect == :CliError
+              puts '         CliError'
+
+              # This set of parameters produces a CLI error
+              assert_raises(Cisco::CliError,
+                            "Assert 'cli error' failed for: #{test}, #{asn}, #{vrf}, #{af}") do
+                bgp_af.send("#{test}=", 'foo') # *** BT: Don't like this ***
```

Comment:
Agreed. Would it make more sense to try one (or all) of the values in `test_values` instead?